Miscellaneous changes

eklimo opened the pull request.

eklimo committed:
Build parameter declarations

eklimo committed:
Build annotation modifiers

eklimo committed:
Implement `ASTModifierNode` methods

eklimo committed:
Implement `ASTUserInterface.getSuperInterfaceName`

eklimo committed:
Generate `ASTField` nodes

eklimo committed:
Implement `ApexQualifiedName.getOperationString`

eklimo committed:
Implement `ASTLiteralExpression.getName`

eklimo committed:
Build enum declarations

eklimo committed:
Build trigger declarations

eklimo committed:
Build trigger variable expressions

eklimo committed:
Implement CPD `ApexTokenizer.tokenize`

eklimo committed:
Add utility function to print `Expression`

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTAnnotation.java:
```diff
@@ -4,20 +4,52 @@
+public class ASTAnnotation extends AbstractApexNode.Single<AnnotationModifier> {
+
+    private static final String[] VALID_ANNOTATION_NAMES = {
+            "SfdcOnly",
```
Unless there's a good reason to order these as such, I recommend alphabetical (otherwise add a code comment for what the order is)

eklimo replied:
The ordering was taken from the Jorje file that defined these annotations. I will sort them, thank you

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTAnnotation.java:
```diff
@@ -4,20 +4,52 @@
+
+public class ASTAnnotation extends AbstractApexNode.Single<AnnotationModifier> {
+
+    private static final String[] VALID_ANNOTATION_NAMES = {
```
wouldn't a Set be a better data structure for this?

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTAnnotation.java:
```diff
@@ -57,8 +87,6 @@ public boolean suppresses(Rule rule) {
-        // return node.getType().isResolved();
-        // TODO(b/239648780)
-        return false;
+        return Arrays.stream(VALID_ANNOTATION_NAMES).anyMatch(name -> node.getName().getString().equalsIgnoreCase(name));
```
instead of using `equalsIgnoreCase` consider the idea of having `VALID_ANNOTATION_NAMES` as all lowercase (which I suggest above be a set) and then see if it contains `name.toLowercase()`

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTModifierNode.java:
```diff
@@ -23,125 +28,133 @@ public Object jjtAccept(ApexParserVisitor visitor, Object data) {
 
+    private static final Map<Keyword, Integer> opcodes = new HashMap<>();
+
+    static {
```
remind me -- we can't depend on Guava, right?

eklimo replied:
PMD does have an existing Guava dependency. How do you recommend that I use it here?

zorzella replied:
Guava has proper support for ImmutableMap, which is a good choice in all sorts of situations, and would make this a more constant-y constant. It also has a syntax for a one-line declaration/assignment which would eliminate the `static` block here.

eklimo committed:
Refactor `ASTAnnotation`

- Sort annotation names
- Make names lowercase
- Use set for names

eklimo committed:
Refactor `ASTModifierNode`

- Use ImmutableMap

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTLiteralExpression.java:
```diff
@@ -87,36 +84,15 @@ public boolean isNull() {
-                    return null;
-                }
-            }).map(Identifier::getValue).orElse(null);
+        if (getParent() instanceof ASTAssignmentExpression && getParent().getParent() instanceof ASTNewKeyValueObjectExpression) {
```
Since you reversed engineered this special case, could you put a comment about it?  (Even if this is as simple as  "given Apex code XXX this returns YYY instead of ZZZ".)

eklimo replied:
Could you clarify what you mean by "reverse engineered"? This code is just the Summit-AST equivalent of the previously commented-out code.

aaronhurst-google replied:
It's not obvious to me that they are equivalent. Are you asserting that every Summit input must produce the same output as the Jorje original, for each possible execution path? Or can you say something higher-level about what the intention is, that you are preserving?

aaronhurst-google replied:
(By "reverse engineered" I was assuming you made a guess about the "intention" and attempted to preserve it. LMK if that wasn't the case.)

eklimo replied:
`ASTNewKeyValueObjectExpression` is a node representing an object constructor with named arguments, e.g.:
```apex
new X(a = 1, b = 2, c = 3)
```
(Only) In this scenario, a literal has the concept of a "name", which this `getName` method is exposing. It seems that Jorje does not build nodes for any of the labels, forcing the previous implementation to use reflection, but my version of the tree builder instead builds the named arguments as an `AssignmentExpression` that contains both the value and the label, making the implementation simpler.

P.S. Although any expression after the label could have a "name", Jorje only has this property for literals.

aaronhurst-google replied:
Super: this example and description are exactly what I would hope for in a comment. (Feel free to leave out anything about Jorje if you want to make it forward-looking.)

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTAnnotation.java:
```diff
@@ -4,20 +4,53 @@
+public class ASTAnnotation extends AbstractApexNode.Single<AnnotationModifier> {
+
+    private static final ImmutableSet<String> VALID_ANNOTATION_NAMES = ImmutableSet.of(
+            "allowcertifiedapex",
```
Did these come from Summit or somewhere on the web?  Would be good to add a comment with a link if possible.

eklimo replied:
They were extracted from a decompiled Jorje class, but most (not all) can be found on this [official page](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation.htm). The missing annotations are `AllowCertifiedApex`, `HiddenFromDoc`, `NamespaceGuard`, `PermGuard`, `PrivateApi`, `SfdcOnly`, `UseConnectDeserializer`, `UseConnectSerializer`, and `VisibleApiVersion`, which are also recognized by Jorje. How do you recommend this to be documented in this case?

aaronhurst-google replied:
I would put exactly that in the comment (link to official page + a callout about the additional cases that exist for backward compatibility).

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTModifierNode.java:
```diff
@@ -25,123 +40,120 @@ public Object jjtAccept(ApexParserVisitor visitor, Object data) {
+                .map(mod -> (KeywordModifier) mod)
+                .filter(mod -> opcodes.containsKey(mod.getKeyword()))
+                .mapToInt(mod -> opcodes.get(mod.getKeyword()))
+                .reduce(0, (current, bit) -> current | bit);
```
Nice stream-based implementation!

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java:
```diff
@@ -144,4 +146,17 @@ public String getNamespace() {
         return null;
     }
+
+    /** Returns the string value of the {@link Expression expr}. */
```
Without looking at the implementation, I would have assumed that this generates a meaningful String for any sort of Expression. But does it exist to handle a specific case?

I would suggest a clearer name (valueIfStringOrNullLiteral?) and/or more explanation in the javadoc.

eklimo replied:
I originally wanted this function to take all `Expression`s as an input but only produce non-`null` values for `LiteralExpression`s, but I decided to only handle `LiteralExpression`s: [`84342a9`](https://github.com/eklimo/pmd/pull/11/commits/84342a9fc9e51b9ee68d09802da50e477305f815). Please let me know if this is better.

aaronhurst-google replied:
Looks good, thanks for the update.

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java:
```diff
@@ -144,4 +146,17 @@ public String getNamespace() {
+            }
+            return ((LiteralExpression) expr).asCodeString();
+        }
+        return null;
```
Prefer `Optional<String>` over `@Nullable String` over `String`

go/java-practices/null

eklimo replied:
~~This style (`return null`) is used all over PMD, so I got used to it, but since this is my own unrelated method, I will use `Optional`. Thank you~~

Irrelevant due to other change

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -643,6 +680,45 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
         ASTModifierNode(modifiers).apply { modifiers.forEach { buildAndSetParent(it, parent = this) } }
 
+    /** Generates additional nodes for the [root] node. */
+    private fun generateAdditional(root: ApexNode<*>) {
```
Suggest more descriptive/specific name

eklimo replied:
```kotlin
/** Generates additional nodes under the [root] node in a second pass after the initial [construction][build] of the AST. */
private fun generateAdditionalNodes(root: ApexNode<*>) {
```
Is this more clear? I seem to have trouble documenting these kinds of functions..

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -643,6 +680,45 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+        findDescendants(root, nodeType = ASTProperty::class).forEach { node -> generateFields(node) }
+    }
+
+    /** Returns all descendants of [root] of type [nodeType], including [root]. */
```
Would this be a useful utility on `Node`?

eklimo replied:
This function operates on `ApexNode`s. Could you elaborate?

aaronhurst-google replied:
n/m. I misread this.

w.r.t. the naming: since this also seeks to `findDescendantsOfType` like the method it wraps, I'm inclined to think it should share that name. But because it behaves differently by including the the root, that distinction seems like it should be capture in the name. any suggestions?

eklimo replied:
Maybe this function can be removed and its uses replaced with the regular `findDescendantsOfType`?

I originally kept the "include-self" behavior for nodes like `ApexRootNode`, where it's more likely that this would be useful (e.g. when searching for `nodeType = ASTUserClass`, the root node itself could be one). This was used to generate additional methods for all `ApexRootNode`s, but currently is unimplemented because it seems to impact no rules. With the  current additional generation behaviors, it's impossible for the root node to be included.

aaronhurst-google reviewed (approved):
LGTM, but see comments

eklimo committed:
Refactor `expressionToString`

- Handle only `LiteralExpression`s; rename to `literalToString`
- Manually discard other `Expression`s at each client

eklimo committed:
Add documentation for `VALID_ANNOTATION_NAMES`

eklimo committed:
Add documentation for `getName`
